### PR TITLE
[Merged by Bors] - feat(Analysis/NormedSpace/FunctionSeries): add eventually versions

### PR DESCRIPTION
**Pull request by CBirkbeck**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 1]
```diff
@@ -5,6 +5,7 @@ Authors: Sébastien Gouëzel
 -/
 import Mathlib.Analysis.Normed.Group.InfiniteSum
 import Mathlib.Topology.Instances.ENNReal
+import Mathlib.Data.Finset.Basic
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:8]: Please check if you need this import explicitly; the bot suggests you don't. If not, please remove.

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 2]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  apply tendsto_finset_range.eventually (tendstoUniformlyOn_tsum_eventually hu ?_ v hv)
+  simp only [eventually_atTop, ge_iff_le] at hfu
+  obtain ⟨N, hN⟩ := hfu
+  use Finset.range N
+  intro b hb x hx n hn
+  apply hN n _ x hx
+  by_contra h
+  simp only [not_le] at h
+  rw [← @Finset.mem_range] at h
+  exact hn (hb h )
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:98]:
```suggestion
  exact hn (hb h)
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 3]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  intro v hv
+  apply tendsto_finset_range.eventually (tendstoUniformlyOn_tsum_eventually hu ?_ v hv)
+  simp only [eventually_atTop, ge_iff_le] at hfu
+  obtain ⟨N, hN⟩ := hfu
+  use Finset.range N
+  intro b hb x hx n hn
+  apply hN n _ x hx
+  by_contra h
+  simp only [not_le] at h
+  rw [← @Finset.mem_range] at h
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:97]:
```suggestion
  rw [← Finset.mem_range] at h
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 4]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  have ht2 := ht (n) (by exact Finset.union_subset_right hn)
+  apply lt_of_le_of_lt _ ht2
+  apply (norm_tsum_le_tsum_norm ?_).trans
+  · apply tsum_le_tsum
+    · intro i
+      apply hN2
+      apply i.2
+    · apply (A.subtype _)
+    · apply (hu.subtype _)
+  · apply (A.subtype _)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:79]:
```suggestion
    · exact A.subtype _
    · exact hu.subtype _
  · exact A.subtype _
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 5]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
       s :=
   fun v hv => tendsto_finset_range.eventually (tendstoUniformlyOn_tsum hu hfu v hv)
 
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version relative to a set, with general index set. -/
+theorem tendstoUniformlyOn_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) {s : Set β}
+    (hfu : ∃ a, ∀ (b : Finset ι), a ⊆ b → ∀ x, x ∈ s → ∀ n, n ∉ b → ‖f n x‖ ≤ u n) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:52]:
```suggestion
    (hfu : ∃ a : Finset ι, ∀ᵉ (b ⊇ a) (x ∈ s) (n ∉ b), ‖f n x‖ ≤ u n) :
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 6]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  use N ∪ t
+  intro n hn x hx
+  have A : Summable fun n => ‖f n x‖ := by
+    apply Summable.add_compl (s := N) Summable.of_finite
+    apply Summable.of_nonneg_of_le (fun _ ↦ norm_nonneg _) _ (hu.subtype _)
+    simp only [comp_apply, Subtype.forall, Set.mem_compl_iff, Finset.mem_coe]
+    apply hN N (by simp only [subset_refl]) x hx
+  rw [dist_eq_norm, ← sum_add_tsum_subtype_compl A.of_norm (n), add_sub_cancel_left]
+  have hN2 := hN (n) (by exact Finset.union_subset_left hn) x hx
+  have ht2 := ht (n) (by exact Finset.union_subset_right hn)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:70]:
```suggestion
  have hN2 := hN n (Finset.union_subset_left hn) x hx
  have ht2 := ht n (Finset.union_subset_right hn)
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 7]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  simp only [gt_iff_lt, eventually_atTop, ge_iff_le, Finset.le_eq_subset] at *
+  obtain ⟨t, ht⟩ := this
+  obtain ⟨N, hN⟩ := hfu
+  use N ∪ t
+  intro n hn x hx
+  have A : Summable fun n => ‖f n x‖ := by
+    apply Summable.add_compl (s := N) Summable.of_finite
+    apply Summable.of_nonneg_of_le (fun _ ↦ norm_nonneg _) _ (hu.subtype _)
+    simp only [comp_apply, Subtype.forall, Set.mem_compl_iff, Finset.mem_coe]
+    apply hN N (by simp only [subset_refl]) x hx
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:67]:
```suggestion
    exact hN N subset_rfl x hx
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 8]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+    (fun x => ∑' n, f n x) atTop s:= by
+  intro v hv
+  apply tendsto_finset_range.eventually (tendstoUniformlyOn_tsum_eventually hu ?_ v hv)
+  simp only [eventually_atTop, ge_iff_le] at hfu
+  obtain ⟨N, hN⟩ := hfu
+  use Finset.range N
+  intro b hb x hx n hn
+  apply hN n _ x hx
+  by_contra h
+  simp only [not_le] at h
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:96]:
```suggestion
  by_contra! h
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 9]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  rw [← @Finset.mem_range] at h
+  exact hn (hb h )
+
 /-- An infinite sum of functions with summable sup norm is the uniform limit of its partial sums.
 Version with general index set. -/
 theorem tendstoUniformly_tsum {f : α → β → F} (hu : Summable u) (hfu : ∀ n x, ‖f n x‖ ≤ u n) :
     TendstoUniformly (fun t : Finset α => fun x => ∑ n ∈ t, f n x)
-      (fun x => ∑' n, f n x) atTop := by
+    (fun x => ∑' n, f n x) atTop := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:104]:
```suggestion
      (fun x => ∑' n, f n x) atTop := by
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 10]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  obtain ⟨t, ht⟩ := this
+  obtain ⟨N, hN⟩ := hfu
+  use N ∪ t
+  intro n hn x hx
+  have A : Summable fun n => ‖f n x‖ := by
+    apply Summable.add_compl (s := N) Summable.of_finite
+    apply Summable.of_nonneg_of_le (fun _ ↦ norm_nonneg _) _ (hu.subtype _)
+    simp only [comp_apply, Subtype.forall, Set.mem_compl_iff, Finset.mem_coe]
+    apply hN N (by simp only [subset_refl]) x hx
+  rw [dist_eq_norm, ← sum_add_tsum_subtype_compl A.of_norm (n), add_sub_cancel_left]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:68]:
```suggestion
  rw [dist_eq_norm, ← sum_add_tsum_subtype_compl A.of_norm n, add_sub_cancel_left]
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 11]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  have hN2 := hN (n) (by exact Finset.union_subset_left hn) x hx
+  have ht2 := ht (n) (by exact Finset.union_subset_right hn)
+  apply lt_of_le_of_lt _ ht2
+  apply (norm_tsum_le_tsum_norm ?_).trans
+  · apply tsum_le_tsum
+    · intro i
+      apply hN2
+      apply i.2
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:76]:
```suggestion
      exact hN2 _ i.2
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 12]
```diff
@@ -59,6 +112,26 @@ theorem tendstoUniformly_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu :
...
+  rw [← tendstoUniformlyOn_univ]
+  apply tendstoUniformlyOn_tsum_eventually hu
+  simpa using hfu
+
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version with index set `ℕ`. -/
+theorem tendstoUniformly_tsum_net_eventually {f : ℕ → α → F} {u : ℕ → ℝ}
+    (hu : Summable u) (hfu : ∀ᶠ n in atTop, ∀ x, ‖f n x‖ ≤ u n) :
+    TendstoUniformly (fun N => fun x => ∑ n ∈ Finset.range N, f n x)
+    (fun x => ∑' n, f n x) atTop := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:130]:
```suggestion
    TendstoUniformly (fun N x => ∑ n ∈ Finset.range N, f n x) (fun x => ∑' n, f n x) atTop := by
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 13]
```diff
@@ -59,6 +112,26 @@ theorem tendstoUniformly_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu :
       atTop :=
   fun v hv => tendsto_finset_range.eventually (tendstoUniformly_tsum hu hfu v hv)
 
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version with general index set. -/
+theorem tendstoUniformly_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) (hfu : ∃ a, ∀ (b : Finset ι), a ⊆ b → ∀ x n, n ∉ b → ‖f n x‖ ≤ u n) :
+    TendstoUniformly (fun t : Finset ι => fun x => ∑ n ∈ t, f n x)
+    (fun x => ∑' n, f n x) atTop := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:120]:
```suggestion
    TendstoUniformly (fun t x => ∑ n ∈ t, f n x) (fun x => ∑' n, f n x) atTop := by
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 14]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
       s :=
   fun v hv => tendsto_finset_range.eventually (tendstoUniformlyOn_tsum hu hfu v hv)
 
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version relative to a set, with general index set. -/
+theorem tendstoUniformlyOn_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) {s : Set β}
+    (hfu : ∃ a, ∀ (b : Finset ι), a ⊆ b → ∀ x, x ∈ s → ∀ n, n ∉ b → ‖f n x‖ ≤ u n) :
+    TendstoUniformlyOn (fun t : Finset ι => fun x => ∑ n ∈ t, f n x)
+    (fun x => ∑' n, f n x) atTop s := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:54]:
```suggestion
    TendstoUniformlyOn (fun t x => ∑ n ∈ t, f n x) (fun x => ∑' n, f n x) atTop s := by
```

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 15]
```diff
@@ -44,11 +45,63 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+    · apply (hu.subtype _)
+  · apply (A.subtype _)
+
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version relative to a set, with index set `ℕ`. -/
+theorem tendstoUniformlyOn_tsum_nat_eventually {α F : Type*} [NormedAddCommGroup F]
+    [CompleteSpace F] {f : ℕ → α → F} {u : ℕ → ℝ} (hu : Summable u) {s : Set α}
+    (hfu : ∀ᶠ n in atTop, ∀ x, x ∈ s → ‖f n x‖ ≤ u n) :
+    TendstoUniformlyOn (fun N => fun x => ∑ n ∈ Finset.range N, f n x)
+    (fun x => ∑' n, f n x) atTop s:= by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:87]:
```suggestion
    TendstoUniformlyOn (fun N x => ∑ n ∈ Finset.range N, f n x) (fun x => ∑' n, f n x) atTop s := by
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
Thanks! Stylistic comments below

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 16]
```diff
@@ -44,6 +44,55 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+    · intro i
+      apply hN2
+      apply i.2
+    · exact (A.subtype _)
+    · exact (hu.subtype _)
+  · exact (A.subtype _)
+
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version relative to a set, with index set `ℕ`. -/
+theorem tendstoUniformlyOn_tsum_nat_eventually {α F : Type*} [NormedAddCommGroup F]
```
Comment (YaelDillies) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:81]: CAn you multiplicativise?

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 17]
```diff
@@ -59,6 +108,24 @@ theorem tendstoUniformly_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu :
...
+theorem tendstoUniformly_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) (hfu : ∃ a, ∀ (b : Finset ι), a ⊆ b → ∀ x n, n ∉ b → ‖f n x‖ ≤ u n) :
+    TendstoUniformly (fun t x => ∑ n ∈ t, f n x) (fun x => ∑' n, f n x) atTop := by
+  rw [← tendstoUniformlyOn_univ]
+  apply tendstoUniformlyOn_tsum_eventually hu
+  simpa using hfu
+
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version with index set `ℕ`. -/
+theorem tendstoUniformly_tsum_net_eventually {f : ℕ → α → F} {u : ℕ → ℝ}
```
Comment (YaelDillies) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:122]:
```suggestion
theorem tendstoUniformly_tsum_nat_eventually {f : ℕ → α → F} {u : ℕ → ℝ}
```
I think it's explicitly not a net ;)

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean (hunk 16 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/NormedSpace/FunctionSeries.lean:81]: So like `tendstoUniformlyOn_tsum ` it depends on `norm_tsum_le_tsum_norm`, which I also dont think we have a multiplicative version of (I could be wrong). I think adding these versions (where possible) would go outside the scope of this PR.

Comment (YaelDillies) [reply Mathlib/Analysis/NormedSpace/FunctionSeries.lean:73]: Hmm okay 😦

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 18]
```diff
@@ -44,6 +44,46 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
...
+  have A : Summable fun n => ‖f n x‖ := by
+    apply Summable.add_compl (s := N) Summable.of_finite
+    apply Summable.of_nonneg_of_le (fun _ ↦ norm_nonneg _) _ (hu.subtype _)
+    simp only [comp_apply, Subtype.forall, Set.mem_compl_iff, Finset.mem_coe]
+    exact hN N subset_rfl x hx
+  rw [dist_eq_norm, ← sum_add_tsum_subtype_compl A.of_norm n, add_sub_cancel_left]
+  apply lt_of_le_of_lt _ (ht n (Finset.union_subset_right hn))
+  apply (norm_tsum_le_tsum_norm (A.subtype _)).trans
+  apply tsum_le_tsum _ (A.subtype _) (hu.subtype _)
+  exact fun i ↦ hN n (Finset.union_subset_left hn) x hx (↑i) i.property
```
Comment (YaelDillies) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:69]: This would probably be a lot more readable as a `calc` block

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer delegate

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 19]
```diff
@@ -44,6 +44,46 @@ theorem tendstoUniformlyOn_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu
       s :=
   fun v hv => tendsto_finset_range.eventually (tendstoUniformlyOn_tsum hu hfu v hv)
 
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version relative to a set, with general index set. -/
+theorem tendstoUniformlyOn_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) {s : Set β}
+    (hfu : ∃ a : Finset ι, ∀ᵉ (b ⊇ a) (x ∈ s) (n ∉ b), ‖f n x‖ ≤ u n) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:51]: Isn't this equivalent to:
```suggestion
theorem tendstoUniformlyOn_tsum_of_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
    (hu : Summable u) {s : Set β}
    (hfu : ∀ᶠ n in cofinite ι, ∀ x ∈ s, ‖f n x‖ ≤ u n) :
```
which is an *actual* `Filter.Eventually` statement? I think this is nicer, the condition should be easier to verify for downstream users, and I bet the proof can be simplified as a result (although I'm not sure about this last claim). Note also the name should be changed.

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 20]
```diff
@@ -59,6 +99,24 @@ theorem tendstoUniformly_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu :
       atTop :=
   fun v hv => tendsto_finset_range.eventually (tendstoUniformly_tsum hu hfu v hv)
 
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version with general index set. -/
+theorem tendstoUniformly_tsum_eventually {ι : Type*} {f : ι → β → F} {u : ι → ℝ}
+    (hu : Summable u) (hfu : ∃ a, ∀ (b : Finset ι), a ⊆ b → ∀ x n, n ∉ b → ‖f n x‖ ≤ u n) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:105]: This should also use `cofinite` and "of" in the name.

File: Mathlib/Analysis/NormedSpace/FunctionSeries.lean [hunk 21]
```diff
@@ -59,6 +99,24 @@ theorem tendstoUniformly_tsum_nat {f : ℕ → β → F} {u : ℕ → ℝ} (hu :
...
+    TendstoUniformly (fun t x => ∑ n ∈ t, f n x) (fun x => ∑' n, f n x) atTop := by
+  rw [← tendstoUniformlyOn_univ]
+  apply tendstoUniformlyOn_tsum_eventually hu
+  simpa using hfu
+
+/-- An infinite sum of functions with eventually summable sup norm is the uniform limit of its
+partial sums. Version with index set `ℕ`. -/
+theorem tendstoUniformly_tsum_nat_eventually {f : ℕ → α → F} {u : ℕ → ℝ}
+    (hu : Summable u) (hfu : ∀ᶠ n in atTop, ∀ x, ‖f n x‖ ≤ u n) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FunctionSeries.lean:114]: I think by reformulating the previous ones, this is not so necessary anymore, as the user can just swap `cofinite ℕ` for `atTop` whenever they wish.

**Review by jcommelin — COMMENTED** [review]
bors merge

**Pull request closed without merging**